shell/platform/linux/fl_application.cc

```diff
@@ -0,0 +1,189 @@
+
+  // Strip out the first argument as it is the binary name.
+  priv->dart_entrypoint_arguments = g_strdupv(*arguments + 1);
+
+  g_autoptr(GError) error = nullptr;
+  if (!g_application_register(application, nullptr, &error)) {
+     g_warning("Failed to register: %s", error->message);
+     *exit_status = 1;
+     return TRUE;
+  }
+
+  // This will only run on the primary instance or this instance with
+  // G_APPLICATION_NON_UNIQUE
+  g_application_activate(application);
+  *exit_status = 0;
+
+  return TRUE;
+}
+
+// Implements GObject::dispose.
+static void fl_application_dispose(GObject* object) {
+  FlApplication* self = FL_APPLICATION(object);
+  FlApplicationPrivate* priv =
+      FL_APPLICATION_PRIVATE(fl_application_get_instance_private(self));
+
+  g_clear_pointer(&priv->dart_entrypoint_arguments, g_strfreev);
+  
+  G_OBJECT_CLASS(fl_application_parent_class)->dispose(object);
+}
+
+static void fl_application_class_init(FlApplicationClass* klass) {
+  G_APPLICATION_CLASS(klass)->activate = fl_application_activate;
+  G_APPLICATION_CLASS(klass)->local_command_line = fl_application_local_command_line;
+  G_OBJECT_CLASS(klass)->dispose = fl_application_dispose;
+
+  /**
+   * FlApplication::view-constructed:
+   * @application: the #FlApplication which emitted the signal
+   * @window: the created #GtkWindow with the FlView embedded
+   * @view: the created #FlView
+   * 
+   * Emitted when a FlView with its GtkWindow is created by FlApplication.
+   * Applications can connect to this signal to register plugins and setup 
+   * communication with the Flutter Engine.
+   * 
+   * Also see #fl_application_connect_view_constructed for a more type-safe
+   * g_signal_connect wrapper for this signal.
+   */
+  signals[VIEW_CONSTRUCTED] = 
+    g_signal_new ("view-constructed",
+                 G_TYPE_FROM_CLASS (klass),
+                 static_cast<GSignalFlags>(G_SIGNAL_RUN_FIRST | G_SIGNAL_NO_RECURSE | G_SIGNAL_NO_HOOKS),
+                 0 /* closure */,
+                 nullptr /* accumulator */,
+                 nullptr /* accumulator data */,
+                 nullptr /* C marshaller */,
+                 G_TYPE_NONE /* return_type */,
+                 2 /* n_params */,
+                 /* param_types */
+                 GTK_TYPE_WINDOW,
+                  fl_view_get_type());
+}
+
+static void fl_application_init(FlApplication* self) {
+  FlApplicationPrivate* priv =
+      FL_APPLICATION_PRIVATE(fl_application_get_instance_private(self));
+
+  priv->activated = FALSE;
+  priv->dart_entrypoint_arguments = nullptr;
+}
+
+G_MODULE_EXPORT
+FlApplication* fl_application_new(const gchar* application_id,
+                                  GApplicationFlags flags) {
+  return FL_APPLICATION(g_object_new(fl_application_get_type(),
+                                     "application-id", application_id,
+                                     "flags", flags,
+                                     nullptr));
+}
+
+G_MODULE_EXPORT
```

Comment:
While this is good in that it adds some type safety, it doesn't match GTK style so I would remove it.